Add `cupy.intersect1d` API.

khushi-411 opened the pull request:
Hi,
The PR aims to add `cupy.intersect1d` API. It follows #6078

Looking forward to your viewpoints :)
Thanks!

emcastillo commented on cupy/_logic/truth.py:
```diff
@@ -98,6 +98,72 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+    Returns the sorted, unique values that are in both of the input arrays.
+
+    Parameters
+    ----------
```
Please, keep the docstring format consistent with the code base

emcastillo replied:
Sorry, it seems that numpy docstrings style is actually ok. https://github.com/cupy/cupy/issues/5437#issuecomment-879148251
My bad! 😅

thanks to @grlee77 for pointing this out to me!

khushi-411 replied:
No worries :v: 

I think it will be good to update other functions too. May I go for it?

asi1024 commented on tests/cupy_tests/logic_tests/test_truth.py:
```diff
@@ -125,6 +125,40 @@ def test(self, xp, dtype):
+    def test_return_indices(self, xp, dtype):
+        a = xp.array([2, 3, 4, 1, 9, 4], dtype=dtype)
+        b = xp.array([7, 5, 1, 2, 9, 3], dtype=dtype)
+        c, i1, i2 = xp.intersect1d(a, b, return_indices=True)
+        ui1 = xp.unravel_index(i1, a.shape)
+        ui2 = xp.unravel_index(i2, b.shape)
+        return c, a[ui1], b[ui2]
```
According to [numPy.intersect1d documentation](https://numpy.org/doc/stable/reference/generated/numpy.intersect1d.html), the first instance of a value is used if there are multiple.
```suggestion
        return xp.intersect1d(a, b, return_indices=True)
```

asi1024 replied:
Could you add a new test case for this check?

khushi-411 replied:
Ah, yes! I missed it. Thanks for pointing out :)

asi1024 replied:
Could you fix also this line? https://github.com/cupy/cupy/pull/6402/files#r795418783

khushi-411 replied:
Yes, done :+1:

amanchhaparia commented on cupy/_logic/truth.py:
```diff
@@ -98,26 +120,99 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+    else:
+        aux.sort()
+
+    mask = aux[1:] == aux[:-1]
```
Who about using the searchsorted as mentioned in https://github.com/cupy/cupy/pull/4018#discussion_r495790724, 

I had also implemented it in PR #6407.

asi1024 replied:
There seems to be not significant difference in performance.

- benchmark:
```py
import cupy
from cupyx.profiler import benchmark


size = (16, 4096, 4096)
scale = 100000000
x1 = cupy.testing.shaped_random(size, scale=scale, dtype='l', seed=0)
x2 = cupy.testing.shaped_random(size, scale=scale, dtype='l', seed=1)

for return_indices in [False, True]:
    perf = benchmark(
        cupy.intersect1d,
        (x1, x2),
        {'return_indices': return_indices},
        n_repeat=100,
    )
    print(perf)
```

- sorting
```
intersect1d         :    CPU:111609.020 us   +/-270.427 (min:111209.120 / max:113111.656) us     GPU-0:113952.656 us   +/-266.501 (min:113556.480 / max:115440.643) us
intersect1d         :    CPU:264716.717 us   +/-128.043 (min:264499.548 / max:265360.571) us     GPU-0:271108.404 us   +/-104.410 (min:270897.156 / max:271538.177) us
```

- searchsorted
```
intersect1d         :    CPU:109051.820 us   +/-196.413 (min:108677.675 / max:109495.282) us     GPU-0:110463.191 us   +/-195.613 (min:110091.263 / max:110893.059) us
intersect1d         :    CPU:244261.040 us   +/-121.246 (min:244080.101 / max:244808.975) us     GPU-0:251270.912 us   +/-112.754 (min:251099.136 / max:251734.009) us
```

kmaehashi replied:
`searchsorted` seems 3~5% better than `sorting`? Why not use it if it's already implemented? 😄

asi1024 commented:
/test mini

kmaehashi commented on cupy/_logic/truth.py:
```diff
@@ -7,18 +7,26 @@
-            size one.
+    Parameters
+    ----------
+    a : cupy.ndarray)
```
```suggestion
    a : cupy.ndarray
```

kmaehashi commented on cupy/_logic/truth.py:
```diff
@@ -7,18 +7,26 @@
+    ----------
+    a : cupy.ndarray)
+        Input array.
+    axis : int or tuple of ints)
```
```suggestion
    axis : int or tuple of ints
```

kmaehashi reviewed (commented):
Made a few nitpicking

khushi-411 commented:
Hi, @asi1024! Hi, @kmaehashi! Hi, @amanchhaparia!
Thanks a lot for reviewing. I've addressed all the reviews. (Apologies for the delay).
A gentle ping to you. Thanks

asi1024 commented:
/test mini

asi1024 commented:
@khushi-411 LGTM. Thanks for your PR!
